plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitSource.java

```diff
@@ -425,16 +423,20 @@ else if (maxSplitBytes * 2 >= remainingBlockBytes) {
             estimatedSplitSizeInBytes.addAndGet(-removedEstimatedSizeInBytes);
             bufferedInternalSplitCount.addAndGet(-removedSplitCount);
 
-            List<InternalHiveSplit> splitsToInsert = splitsToInsertBuilder.build();
-            List<ConnectorSplit> result = resultBuilder.build();
-            return new AsyncQueue.BorrowResult<>(splitsToInsert, result);
+            return new AsyncQueue.BorrowResult<>(splitsToInsertBuilder.build(), resultBuilder.build());
         });
 
-        ListenableFuture<ConnectorSplitBatch> transform = Futures.transform(future, splits -> {
-            requireNonNull(splits, "splits is null");
+        return toCompletableFuture(future).thenApply(hiveSplits -> {
+            requireNonNull(hiveSplits, "hiveSplits is null");
             if (recordScannedFiles) {
-                splits.forEach(split -> scannedFilePaths.add(((HiveSplit) split).getPath()));
+                hiveSplits.stream()
+                        .filter(split -> split.getStart() == 0)
```

Comment:
is there a chance that split startin at `0` could be filtered out alltogether? While still another splits for same file are present? Can this happen with DF for example? cc: @raunaqmorarka ?

Reply:
It should not be possible, no- the first split generated for a file will always have `start == 0`. Now, it is possible that a file will get eliminated by dynamic filters after the first split is produced which could then eliminate the remainder of the file from being scanned, but that could happen even with the previous implementation. As far as I can tell, this logic is only necessary for "optimizing small files" and collecting the files scanned to delete them after the contents are rewritten- which should not have dynamic filters in the first place because it would be unsafe for this reason with or without this improvement.